.travis.yml

```diff
@@ -25,63 +25,93 @@
 
 env:
   global:
-  - PATH="${TRAVIS_BUILD_DIR}/bin:$PATH"
-  # Only run the generic tests on Travis CI.
-  - CYLC_TEST_RUN_PLATFORM=false
-  # Custom diff command to ignore Xlib errors (xvfb has not RANDR extension).
-  - CYLC_TEST_DIFF_CMD="diff -I Xlib -u"
-  # This coverage RC file is created under the script task
-  - COVERAGE_PROCESS_START="/tmp/.coveragerc"
-  matrix:
-  - CHUNK="1/4"
-  - CHUNK="2/4"
-  - CHUNK="3/4"
-  - CHUNK="4/4"
+    - PATH="${TRAVIS_BUILD_DIR}/bin:$PATH"
+    # Only run the generic tests on Travis CI.
+    - CYLC_TEST_RUN_PLATFORM=false
+    # Custom diff command to ignore Xlib errors (xvfb has not RANDR extension).
+    - CYLC_TEST_DIFF_CMD="diff -I Xlib -u"
+    # This coverage RC file is created under the script task
+    - COVERAGE_PROCESS_START="/tmp/.coveragerc"
+
+unittest: &unittest
+  stage: unit-test
+  install:
+    - .travis/install.sh unit-tests
+  script:
+    - export PYTHONPATH=$(pwd -P)/lib/
+    - pycodestyle
+    - pytest --cov-append --cov=lib/cylc --cov=lib/parsec
+  after_success:
+    # Report metrics, such as coverage
+    - coverage combine --append
+    - coverage xml --ignore-errors
+    - bash <(curl -s https://codecov.io/bash)
 
-# this is the default test stage, which is used for functional-tests
-install: .travis/install.sh functional-tests docs
-script:
-- export PYTHONPATH="${TRAVIS_BUILD_DIR}/.travis"
-# When we run cylc commands, there are processes being forked, that get a
-# new working directory. As .coveragerc contains relatives paths, it fails
-# to produce the correct coverage, unless we use absolute paths. The `sed`
-# call below tries to define the data_file, and sources locations for Travis.
-- sed -e "s|data_file=.coverage|data_file=${TRAVIS_BUILD_DIR}/.coverage|g; s|./bin|${TRAVIS_BUILD_DIR}/bin|g; s|./lib|${TRAVIS_BUILD_DIR}/lib|g" .coveragerc > /tmp/.coveragerc
-# And some tests fail if we touch files in the git working directory, due
-# to Cylc's version appearing with the "dirty" suffix. To avoid this, we
-# are using a new coveragerc created under the temporary directory.
-- coverage run .travis/cover.py
-- unset PYTHONPATH
-after_script: .travis/after_script.sh
-after_success:
-# Report metrics, such as coverage
-- coverage combine --append
-- coverage xml --ignore-errors
-- bash <(curl -s https://codecov.io/bash)
+regressiontest: &regressiontest
+  stage: functional-test
+  python:
+    - "3.7"
+  install:
+    - .travis/install.sh functional-tests docs
+  script:
+    - export PYTHONPATH="${TRAVIS_BUILD_DIR}/.travis"
+    - .travis/coverage-data.sh
+    - coverage run .travis/cover.py
+    - unset PYTHONPATH
+  after_failure:
+    - .travis/after_failure.sh
+  after_success:
+    # Report metrics, such as coverage
+    - coverage combine --append
+    - coverage xml --ignore-errors
+    - bash <(curl -s https://codecov.io/bash)
 
 jobs:
   include:
-  - stage: unit-test
+
+  # run unittests at each supported Python version
+  - <<: *unittest
+    python: "3.6"
```

Comment:
Any reason to support Python 3.6 just in here?

Reply:
We aught to tests the oldest and newest supported Python versions. Running the whole test battery under both Python versions is probably a bit much so what I've started here is:

* Run all tests under the latest Python version
* Run quick tests under the oldest supported Python Version